### PR TITLE
Port ApplicationId

### LINKED ISSUE #12399: Port ApplicationId
```
T:System.ApplicationId
M:System.ApplicationId.#ctor(System.Byte[],System.String,System.Version,System.String,System.String)
P:System.ApplicationId.Culture
M:System.ApplicationId.get_Culture
P:System.ApplicationId.Name
M:System.ApplicationId.get_Name
P:System.ApplicationId.ProcessorArchitecture
M:System.ApplicationId.get_ProcessorArchitecture
P:System.ApplicationId.PublicKeyToken
M:System.ApplicationId.get_PublicKeyToken
P:System.ApplicationId.Version
M:System.ApplicationId.get_Version
M:System.ApplicationId.Copy
M:System.ApplicationId.Equals(System.Object)
M:System.ApplicationId.GetHashCode
M:System.ApplicationId.ToString
```

### PR DESCRIPTION
**Pull request by alexperovich**
Fixes #12399

### CODE REVIEW

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+using System.IO;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Text;
+
+namespace System
+{
+    public sealed class ApplicationId
+    {
+        private byte[] m_publicKeyToken;
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: `m_publicKeyToken` => `_publicKeyToken`. Also, `readonly`?

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+        public ApplicationId(byte[] publicKeyToken, string name, Version version, string processorArchitecture, string culture)
+        {
+            if (name == null) throw new ArgumentNullException(nameof(name));
+            if (name.Length == 0) throw new ArgumentException(SR.Argument_EmptyApplicationName);
+            if (version == null) throw new ArgumentNullException(nameof(version));
+            if (publicKeyToken == null) throw new ArgumentNullException(nameof(publicKeyToken));
+
+            m_publicKeyToken = new byte[publicKeyToken.Length];
+            Array.Copy(publicKeyToken, m_publicKeyToken, publicKeyToken.Length);
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: Nit:

```c#
_publicKeyToken = (byte[])publicKeyToken.Clone();
```

?

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        public Version Version { get; }
+
+        public byte[] PublicKeyToken
+        {
+            get
+            {
+                var result = new byte[m_publicKeyToken.Length];
+                Array.Copy(m_publicKeyToken, result, m_publicKeyToken.Length);
+                return result;
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: Nit:

```c#
public byte[] PublicKeyToken => (byte[])_publicKeyToken.Clone();
```

?

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            }
+        }
+
+        public ApplicationId Copy() => new ApplicationId(m_publicKeyToken, Name, Version, ProcessorArchitecture, Culture);
+
+        public override string ToString ()
+        {
+            StringBuilder sb = StringBuilderCache.Acquire();
+            sb.Append(this.Name);
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: Nit: `this.` isn't necessary

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            if (Culture != null)
+            {
+                sb.Append(", culture=\"");
+                sb.Append(Culture);
+                sb.Append("\"");
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: Nit: `sb.Append('"');`

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                sb.Append(this.ProcessorArchitecture);
+                sb.Append("\"");
+            }
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        private static char HexDigit(int num) =>
+            (char)((num < 10) ? (num + '0') : (num + ('A' - 10)));
+        
+        private static String EncodeHexString(byte[] sArray)
```
Comment (justinvp) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: Nit: `String` -> `string`

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            if (Culture != null)
+            {
+                sb.Append(", culture=\"");
+                sb.Append(Culture);
+                sb.Append("\"");
+            }
+            sb.Append(", version=\"");
+            sb.Append(Version.ToString());
+            sb.Append("\"");
+            if (m_publicKeyToken != null) {
```
Comment (ianhays) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: nit: bracketing style has been up to this point consistently next-line.

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            return result;
+        }
+ 
+        public override bool Equals (Object o)
+        {
+            ApplicationId other = (o as ApplicationId);
+            if (other == null)
+                return false;
+ 
+            if (!(Equals(this.Name, other.Name) &&
```
Comment (ianhays) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: lots of unnecessary "this" scattered throughout.

File: src/System.Runtime.Extensions/src/System/ApplicationId.cs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                  Equals(this.Version, other.Version) &&
+                  Equals(this.ProcessorArchitecture, other.ProcessorArchitecture) &&
+                  Equals(this.Culture, other.Culture)))
+                return false;
+ 
+            if (this.m_publicKeyToken.Length != other.m_publicKeyToken.Length)
+                return false;
+ 
+            for (int i = 0; i < this.m_publicKeyToken.Length; ++i)
```
Comment (ianhays) [inline src/System.Runtime.Extensions/src/System/ApplicationId.cs]: nit: postfix increment is more common in fors across the repo.

**Pull request merged**